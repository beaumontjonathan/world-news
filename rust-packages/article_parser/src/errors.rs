use chrono::ParseError as ChronoParseError;
use htmlescape::DecodeErr;
use scraper::Selector;
use serde_json::Error as SerdeJsonError;

#[derive(Debug)]
pub enum InternalParseError {
  UnknownError(&'static str),
  HtmlStructureError(HtmlStructureError),
  JsonParseError(JsonParseError),
  InvalidDateTime(ChronoParseError),
  HtmlDecodeErr(DecodeErr),
}

#[derive(Debug)]
pub enum JsonParseErrorCause {
  InvalidLdJson,
  InvalidDataAttrJson,
}

#[derive(Debug)]
pub struct JsonParseError {
  pub cause: JsonParseErrorCause,
  pub serde_error: SerdeJsonError,
}

#[derive(Debug)]
pub enum HtmlErrorCause {
  MissingElement,
  NonUniqueElement,
  MissingAttribute(String),
}

#[derive(Debug)]
pub struct HtmlStructureError {
  pub cause: HtmlErrorCause,
  pub selector: &'static Selector,
}

impl InternalParseError {
  pub fn html(error: HtmlStructureError) -> InternalParseError {
    InternalParseError::HtmlStructureError(error)
  }

  pub fn json(error: JsonParseError) -> InternalParseError {
    InternalParseError::JsonParseError(error)
  }

  pub fn ld_json(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidLdJson,
      serde_error,
    })
  }

  pub fn json_data_attr(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidDataAttrJson,
      serde_error,
    })
  }

  pub fn unknown(s: &'static str) -> InternalParseError {
    InternalParseError::UnknownError(s)
  }

  pub fn chrono(err: ChronoParseError) -> InternalParseError {
    InternalParseError::InvalidDateTime(err)
  }

  pub fn decode_html(err: DecodeErr) -> InternalParseError {
    InternalParseError::HtmlDecodeErr(err)
  }
}

impl From<ChronoParseError> for InternalParseError {
  fn from(err: ChronoParseError) -> Self {
    InternalParseError::chrono(err)
  }
}

impl From<DecodeErr> for InternalParseError {
  fn from(err: DecodeErr) -> Self {
    InternalParseError::decode_html(err)
  }
}
