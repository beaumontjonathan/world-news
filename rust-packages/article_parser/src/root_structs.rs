use chrono::{DateTime, Utc};
use serde::Serialize;
use std::fmt::Debug;

#[derive(Debug, Serialize)]
pub struct Image {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Video {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
pub enum Publisher {
  BBC,
  DailyMail,
  Independent,
  Metro,
  Mirror,
  Sky,
  Sun,
  Guardian,
}

use Publisher::*;

impl From<String> for Publisher {
  fn from(publisher: String) -> Self {
    match publisher.to_lowercase().as_str() {
      "bbc" => BBC,
      "dailymail" => DailyMail,
      "independent" => Independent,
      "metro" => Metro,
      "mirror" => Mirror,
      "sky" => Sky,
      "sun" => Sun,
      "theguardian" | "guardian" => Guardian,
      _ => panic!("Invalid publisher type"),
    }
  }
}

#[derive(Debug, Serialize)]
pub struct NewsArticleContent {
  pub headline: String,
  pub twitter_headline: Option<String>,
  pub description: String,
  pub thumbnail: Image,
  pub categories: Vec<String>,
  pub images: Vec<Image>,
  pub videos: Vec<Video>,
  pub body: String,
  pub date_updated: DateTime<Utc>,
  pub date_published: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct VideoArticleContent {
  pub something: String,
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, Serialize)]
pub enum ParsedPageContent {
  NewsArticle(NewsArticleContent),
  VideoArticle(VideoArticleContent),
}

#[derive(Debug, Serialize)]
pub struct PageMeta {
  pub publisher: Publisher,
  pub url: String,
  pub date_parsed: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct ParsedPage {
  pub meta: PageMeta,
  pub content: ParsedPageContent,
}
