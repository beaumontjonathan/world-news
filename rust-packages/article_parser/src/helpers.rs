use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
// use chrono::{DateTime, Utc};
use crate::root_structs::ParsedPageContent;
use lazy_static::lazy_static;
use regex::Regex;
use scraper::{ElementRef, Html, Selector};
// use std::fmt::{self, Debug, Display, Formatter};
pub trait ToOwnedOrNone {
  fn to_owned_or_none(self) -> Option<String>;
}

impl<'a> ToOwnedOrNone for Option<&'a str> {
  fn to_owned_or_none(self) -> Option<String> {
    self.and_then(|mut s| {
      s = s.trim();
      if s.is_empty() {
        None
      } else {
        Some(String::from(s))
      }
    })
  }
}
// pub trait NoneIfEmpty
// where
//   Self: Sized,
// {
//   fn none_if<F>(self, callback: F) -> Option<Self>
//   where
//     F: Fn(&Self) -> bool,
//   {
//     if callback(&self) {
//       None
//     } else {
//       Some(self)
//     }
//   }
// }
// impl<T> NoneIfEmpty for T {}

pub type ParseResult = Result<ParsedPageContent, InternalParseError>;

pub trait FixHtmlWhitespace {
  fn fix_html_whitespace(self) -> Self;
}

fn fix_whitespace(text: String) -> String {
  lazy_static! {
    static ref RE: Regex = Regex::new(r#"\s+"#).unwrap();
  }
  RE.replace_all(&text, " ").into_owned()
}

impl FixHtmlWhitespace for String {
  fn fix_html_whitespace(self) -> String {
    fix_whitespace(self)
  }
}

pub trait AttrValueOrNone {
  fn attr_value_or_none(&self, attr: &str) -> Option<String>;
}

impl<'a> AttrValueOrNone for ElementRef<'a> {
  fn attr_value_or_none(&self, attr: &str) -> Option<String> {
    self.value().attr(attr).to_owned_or_none()
  }
}

pub trait FullFixedHtmlText {
  fn full_fixed_html_text(self) -> String;
}

impl<'a> FullFixedHtmlText for ElementRef<'a> {
  fn full_fixed_html_text(self) -> String {
    self
      .text()
      .map(|text| text.trim().to_string())
      .collect::<Vec<String>>()
      .join(" ")
      .fix_html_whitespace()
      .trim()
      .to_owned()
  }
}

pub trait JoinWithNewline {
  fn join_with_newline(self) -> String;
}

impl<T> JoinWithNewline for T
where
  T: Iterator<Item = String>,
{
  fn join_with_newline(self) -> String {
    self.collect::<Vec<String>>().join("\n").trim().to_owned()
  }
}

pub trait UniqueElementRef<'a> {
  fn unique(self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause>;
}

impl<'a, T> UniqueElementRef<'a> for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn unique(mut self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause> {
    match (self.next(), self.next()) {
      (None, _) => Err(HtmlErrorCause::MissingElement),
      (_, Some(_)) => Err(HtmlErrorCause::NonUniqueElement),
      (Some(el_ref), None) => Ok(el_ref),
    }
  }
}

pub trait SelectHelpers {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError>;
}

impl SelectHelpers for Html {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

impl SelectHelpers for scraper::element_ref::ElementRef<'_> {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

pub trait FormattedText {
  fn formatted_text(self) -> String;
}

impl<'a, T> FormattedText for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn formatted_text(self) -> String {
    self.map(|p| p.full_fixed_html_text()).join_with_newline()
  }
}

pub trait ChildrenHelpers {
  fn children_first(&self) -> Option<ElementRef>;
  fn children_all(&self) -> Vec<ElementRef>;
}

impl ChildrenHelpers for scraper::element_ref::ElementRef<'_> {
  fn children_first(&self) -> Option<ElementRef> {
    self.children_all().into_iter().next()
  }

  fn children_all(&self) -> Vec<ElementRef> {
    self.children().filter_map(ElementRef::wrap).collect()
  }
}
