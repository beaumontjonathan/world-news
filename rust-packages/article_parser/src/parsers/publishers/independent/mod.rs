use crate::errors::InternalParseError;
use crate::helpers::{
  AttrValueOrNone, FullFixedHtmlText, JoinWithNewline, ParseResult, SelectHelpers,
};
use crate::my_selectors;
use crate::root_structs::{Image, NewsArticleContent, ParsedPageContent, Video};
use chrono::DateTime;
use htmlescape::decode_html;
use scraper::{Html, Selector};

my_selectors! {
  ARTICLE_SELECTOR => "article",
  HEADLINE_SELECTOR => ".headline",
  TWITTER_TITLE_SELECTOR => "meta[name=\"twitter:title\"]",
  DESCRIPTION_SELECTOR => ".sub-headline",
  HERO_IMAGE_SELECTOR => ".hero-image",
  IMAGE_WRAPPER_CHILD_SELECTOR => ":scope > .image-wrapper",
  META_OG_IMAGE => "meta[property=\"og:image\"]",
  SECTION_SELECTOR => "meta[name=\"article:section\"]",
  SUBSECTION_SELECTOR => "meta[name=\"article:subsection\"]",
  STORY_BODY_SELECTOR => "div.main-content > div.body-content",
  STORY_BODY_IMAGE_SELECTOR => ":scope > figure:not(.i-gallery):not(.video)",
  AMP_IMG_SELECTOR => "amp-img",
  FIGCAPTION_SELECTOR => "figcaption",
  FIGCAPTION_CAPTION_SELECTOR => "figcaption.caption",
  FIGURE_VIDEO_SELECTOR => ":scope > figure.video",
  P_SELECTOR => ":scope > p",
  DATE_UPDATED => "meta[property=\"article:modified_time\"]",
  DATE_PUBLISHED => "meta[property=\"article:published_time\"]",
}

const FORMAT: &str = "%Y-%m-%dT%H:%M:%S%z";

pub fn parse(document: Html) -> ParseResult {
  // DOCUMENTS
  let story_body = document.select_unique(&STORY_BODY_SELECTOR)?;

  // HEADLINE
  let headline = document
    .select_unique(&HEADLINE_SELECTOR)?
    .full_fixed_html_text();

  // TWITTER HEADLINE
  let twitter_headline = document
    .select_unique_attr(&TWITTER_TITLE_SELECTOR, "content")
    .ok();

  // DESCRIPTION
  let description =
    document
      .select_unique(&DESCRIPTION_SELECTOR)
      .map_or(Ok(String::new()), |el| {
        decode_html(&el.full_fixed_html_text())
          .map_err(|_| InternalParseError::unknown("Unknown error: html parsing"))
      })?;

  // THUMBNAIL
  let thumbnail = {
    if let Some(hero_image) = document.select(&HERO_IMAGE_SELECTOR).next() {
      let hero_image_or_optional_wrapper = hero_image
        .select(&IMAGE_WRAPPER_CHILD_SELECTOR)
        .next()
        .unwrap_or(hero_image);

      let amp_img = hero_image_or_optional_wrapper.select_unique(&AMP_IMG_SELECTOR)?;

      let alt = amp_img.attr_value_or_none("alt");
      let url = amp_img.attr_value_or_none("src");
      let caption = amp_img.attr_value_or_none("title");

      Image { alt, url, caption }
    } else {
      let url = document.select_unique_attr(&META_OG_IMAGE, "content").ok();
      Image {
        alt: None,
        url,
        caption: None,
      }
    }
  };

  // CATEGORIES
  let section_category = document
    .select_unique_attr(&SECTION_SELECTOR, "content")
    .ok()
    .filter(|x| !x.is_empty());
  let subsection_categories = document
    .select_unique_attr(&SUBSECTION_SELECTOR, "content")
    .ok()
    .filter(|x| !x.is_empty());
  let categories = match (section_category, subsection_categories) {
    (Some(category), Some(sub_categories)) => format!("{}, {}", category, sub_categories)
      .split(", ")
      .map(|s| s.to_owned())
      .collect(),
    (Some(category), None) => vec![category],
    (None, Some(sub_categories)) => sub_categories.split(", ").map(|s| s.to_owned()).collect(),
    (None, None) => Vec::new(),
  };

  // IMAGES
  let images = story_body
    .select(&STORY_BODY_IMAGE_SELECTOR)
    // here are some things
    .filter_map(|figure| {
      figure
        .select_unique(&AMP_IMG_SELECTOR)
        .map(|img| {
          let url = img.value().attr("src").map(|s| s.to_owned());
          let caption = figure
            .select(&FIGCAPTION_SELECTOR)
            .next()
            .map(|e| e.full_fixed_html_text());
          Image {
            alt: None,
            url,
            caption,
          }
        })
        .ok()
    })
    .collect::<Vec<_>>();

  // VIDEOS
  let videos = story_body
    .select(&FIGURE_VIDEO_SELECTOR)
    .map(|el| {
      let caption = Some(
        el.select_unique(&FIGCAPTION_CAPTION_SELECTOR)?
          .full_fixed_html_text(),
      );
      Ok(Video {
        alt: None,
        url: None,
        caption,
      })
    })
    .collect::<Result<Vec<_>, InternalParseError>>()?;

  // BODY
  let body = story_body
    .select(&P_SELECTOR)
    .map(|p| p.full_fixed_html_text())
    .join_with_newline();

  // DATE UPDATED
  let date_updated = document.select_unique_attr(&DATE_UPDATED, "content")?;
  let date_updated = DateTime::parse_from_str(&date_updated, FORMAT)?.into();

  //DATE PUBLISHED
  let date_published = document.select_unique_attr(&DATE_PUBLISHED, "content")?;
  let date_published = DateTime::parse_from_str(&date_published, FORMAT)?.into();

  // Parsed::Err("I didn't bother to do this.".to_owned())
  Ok(ParsedPageContent::NewsArticle(NewsArticleContent {
    headline,
    twitter_headline,
    description,
    thumbnail,
    categories,
    images,
    videos,
    body,
    date_updated,
    date_published,
  }))
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::test_helpers::helpers::read_html_doc;
  use insta::assert_debug_snapshot;

  #[test]
  fn test_file_01() {
    let document = read_html_doc("test-01.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_02() {
    let document = read_html_doc("test-02.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_03() {
    let document = read_html_doc("test-03.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_04() {
    let document = read_html_doc("test-04.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_05() {
    let document = read_html_doc("test-05.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }
}
