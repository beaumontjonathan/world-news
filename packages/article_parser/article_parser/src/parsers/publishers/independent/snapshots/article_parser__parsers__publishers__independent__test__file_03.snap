---
source: src/parsers/publishers/independent/mod.rs
expression: article
---
Ok(
    NewsArticle(
        NewsArticleContent {
            headline: "\'Deeply inappropriate\': Boris Johnson facing mounting demands to suspend all post-Brexit trade talks during coronavirus",
            twitter_headline: Some(
                "Boris Johnson facing mounting demands to suspend all post-Brexit tra…",
            ),
            description: "\'We call on the government to pause all trade negotiations until the covid-19 crisis is under control and to inform both the public and potential trade partners of necessary action\'",
            thumbnail: Image {
                alt: None,
                url: Some(
                    "https://static.independent.co.uk/s3fs-public/thumbnails/image/2020/03/26/11/boris-johnson.jpg?w968h681",
                ),
                caption: None,
            },
            categories: [
                "UK Politics",
            ],
            images: [],
            videos: [],
            body: "Boris Johnson is facing escalating pressure to suspend all post- Brexit trade negotiations and concentrate government’s resources on fighting the coronavirus pandemic.\nIn a letter coordinated by the Trade Justice Movement and signed by several charities – including Friends of the Earth and the RSPCA – the prime minister is urged to pause talks until parliament returns and the crisis is under under control.\nAddressed to both Mr Johnson and Liz Truss , the international trade secretary, the groups claim it would be “deeply inappropriate” for trade talks between the UK and the US to commence at a time when both countries have imposed widespread lockdowns.\n“The UK government’s attention should be focused on dealing with the public health and economic crises which face our nation,” they added.\nCiting several “high risk issues”, including food standards, climate action, medicine prices and the NHS , the letter added: “This debate cannot happen amidst national lockdown and with parliament closed.\n“We call on the government to pause all trade negotiations until the covid-19 crisis is under control and to inform both the public and potential trade partners of necessary action.”\nOn Wednesday, the Freight Transport Association, representing UK logistics, also urged the government to urgently seek an extension to the Brexit transition period amid widespread disruption to business across the country.\n“This is not about the relative merits of Brexit, or any trading arrangements which our industry will need to adopt,” said the organisation’s policy director Elizabeth de Jong.\nShe added: “This is purely and simply so the businesses tasked with keeping the UK’s supply chain intact can concentrate on the serious issues which the covid-19 pandemic is placing on industry.\n“Logistics is facing unprecedented challenges, both in terms of keeping the UK economy supplies with all the goods it needs to function, as well as coping with the increased disruption to staffing levels caused by sickness and self-isolation and concerns about the viability of their busies.\n“Our first priority is always to deliver for our customers, and there is simply not enough capacity available to plan the major structural changes needed to implement a successful departure from the EU , as well as the myriad of other planned legislation changes on the horizon, as well as dealing with unprecedented pressures caused by covid-19.”\nAnd on the same day, the Welsh and Scottish governments called on the prime minister to halt the negotiations over the UK-EU’s future relationship, with the latter claiming the talks were an “unwelcome and unnecessary distraction” from the current crisis.\nRoseanna Cunningham, the environment secretary in Scotland, said the outbreak of the coronavirus was a “distressing period” for the entire country, adding: “All available government resource needs to be put towards tackling covid-19 and mitigating its effects on almost every area of Scottish society”.\nShe continued: “In this context, the Scottish government feels that a Brexit Scotland did not vote for, and does not want, is an unwelcome and unnecessary distraction. Continuing to plough on could seriously harm our ability to tackle a virus which threatens lives and livelihoods across these islands.”\nLast week, the EU Commission’s chief negotiator, Michel Barnier, also tested positive for coronavirus and his opposite number in the UK David Frost is also self-isolating after showing “mild” symptoms . But despite the escalating threat of covid-19, Downing Street has repeatedly insisted that there will be no extension to the 11-month transition period.",
            date_updated: 2020-03-26T12:11:50Z,
            date_published: 2020-03-26T11:51:00Z,
        },
    ),
)
